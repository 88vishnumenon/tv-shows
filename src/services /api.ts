import { AxiosResponse } from "axios";
import axios from "../shared/services/interceptor";
import { TVShowDetailValues, TVShowDetails } from "../types/types";



export const  listallTvShows = async() : Promise<TVShowDetailValues[]>=>{
 const results = await axios.get("/shows");
 return results.data;
}

export const  listShowsBySearch= async(searchValue:string) : Promise<TVShowDetails[]>=>{
    const results = await axios.get(`/search/shows?q=${searchValue}`);
    return results.data;
   }