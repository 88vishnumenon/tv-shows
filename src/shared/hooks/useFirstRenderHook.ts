import { useRef, useEffect } from "react";

const useIsFirstRender = () => {
    const is_first_render = useRef(true);
    console.log("useIsFirstRender rendered");

    useEffect(() => {
        is_first_render.current = false;
        console.log("useIsFirstRender changed to false");
    }, []);
    console.log("useIsFirstRender value returned");
    return is_first_render.current;
};
export default useIsFirstRender;