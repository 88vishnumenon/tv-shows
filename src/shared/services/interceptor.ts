import axios, { InternalAxiosRequestConfig } from "axios";

import { AxiosError, AxiosRequestConfig, AxiosResponse } from "axios";

const customAxios = axios.create({
  baseURL: `https://api.tvmaze.com`,
});

const onRequest = (config: InternalAxiosRequestConfig): InternalAxiosRequestConfig => {
  return config;
};

const onRequestError = (error: AxiosError): Promise<AxiosError> => {
  console.error(`[request error] [${JSON.stringify(error)}]`);
  return Promise.reject(error);
};

const onResponse = (response: AxiosResponse): AxiosResponse => {
  return response;
};

const onResponseError = (error: AxiosError): Promise<AxiosError> => {
    // not showing the error when the error is of cancelled request
    // if (error.message != "axios request cancelled"){
    //   toast.error("Oops, something went wrong!");
    // } 
    return Promise.reject(error);
};

customAxios.interceptors.request.use(
  (request) => onRequest(request),
  (error) => onRequestError(error)
);

customAxios.interceptors.response.use(
  (response) => onResponse(response),
  (error) => onResponseError(error)
);

export default customAxios;