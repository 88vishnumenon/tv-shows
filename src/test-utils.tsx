import { Provider } from "react-redux";
import { RootState, getStoreWithState, store } from "./store/store";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import * as React from "react";
import { TVShow } from "./types/types";

export function renderWithContext(
  element: React.ReactElement,
  state?: RootState
) {
  const store = getStoreWithState(state);
  const utils = render(
    <Provider store={store}>
      <Router>{element}</Router>
    </Provider>
  );
  return { store, ...utils };
}

export function getStateWithItems(
  tvShows: TVShow[]
): RootState {
  const state: RootState = {
    tvshow: {
      showList: tvShows,
      generesList: ["DRAMA"],
      searchString: "",
      error: false,
    },
  };

  return state;
}
