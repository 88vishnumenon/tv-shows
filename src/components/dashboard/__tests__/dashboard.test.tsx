import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import * as React from "react";
import * as api  from "../../../services/api";
import * as DebounceUse from "../../../shared/hooks/useDebounceHook";

import Dashboard from "../dashboard";
import { renderWithContext } from "../../../test-utils";



const listShowsBySearchSpy = jest.spyOn(api,"listShowsBySearch")
listShowsBySearchSpy.mockResolvedValue([]);


beforeEach(() => {
    renderWithContext(<Dashboard />);
 } );

describe("dashboard component", () => {
  test("should match snapshot", async () => {
    const { asFragment } = renderWithContext(<Dashboard />);
    expect(asFragment()).toMatchSnapshot();
  });


  it("should render the search bar", () => {
    const element = screen.getByTestId("search-bar");
    expect(element).toBeInTheDocument;
  });

  it("should render the welcome message bar", () => {
    const element = screen.getByTestId("welcome-message");
    expect(element).toBeInTheDocument;
  });

});
