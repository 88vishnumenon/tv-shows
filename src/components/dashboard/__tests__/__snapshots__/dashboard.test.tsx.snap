// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard component should match snapshot 1`] = `
<DocumentFragment>
  <article
    class="dashboardWrapper"
  >
    <section
      class="inputWrapper"
    >
      <input
        class="searchInput"
        data-testid="search-bar"
        placeholder="Search you favourite tv show"
        type="text"
        value=""
      />
    </section>
    <section
      class="noResults"
    >
      <div
        class="noResultsData"
      >
        <img
          class="welcomeImg"
          src="assets/images/welcome.png"
        />
        <span
          class="noDataMessage"
          data-testid="welcome-message"
        >
          You can search for your favourite TV shows here!
        </span>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
